packages/flutter/lib/src/widgets/inherited_model.dart

```diff
@@ -0,0 +1,234 @@
+    if (model == null)
+      return;
+
+    result.add(model);
+
+    assert(model.widget is T);
+    final T modelWidget = model.widget;
+    if (aspect == null || modelWidget.aspects == null || modelWidget.aspects.contains(aspect))
+      return;
+
+    Element modelParent;
+    model.visitAncestorElements((Element ancestor) {
+      modelParent = ancestor;
+      return false;
+    });
+    if (modelParent == null)
+      return;
+    _findModels<T>(modelParent, aspect, result);
+  }
+
+  /// Makes [context] dependent on the specified [aspect] of an [InheritedModel]
+  /// of type T.
+  ///
+  /// When the given [aspect] of the model changes, the context will be
+  /// rebuilt. The [updateShouldNotifyDependent] method must determine if a
+  /// change in the model widget corresponds to an `aspect` value.
+  ///
+  /// The dependency created by this method targets the first
+  /// [InheritedModel] ancestor of type T whose [InheritedModel.aspects] set
+  /// is either null (the default) or contains [aspect]. If [aspect] is
+  /// null then it's the first ancestor of type T.
+  ///
+  /// If [aspect] is null this method is the same as
+  /// `context.inheritFromWidgetOfExactType(T)`.
+  static T inheritFrom<T extends InheritedModel<Object>>(BuildContext context, { Object aspect }) {
+    if (aspect == null)
+      return context.inheritFromWidgetOfExactType(T);
+
+    // Create a dependency on all of the type T ancestor models up until
+    // a model is found whose aspects are null or contain the given aspect.
+    // Rebuilding one of the intermediate models only causes its dependents
+    // to be rebuilt if its aspects property changes (see _shouldNotify).
+    final List<InheritedElement> models = <InheritedElement>[];
+    _findModels<T>(context, aspect, models);
+    if (models.isEmpty)
+      return null;
+    final InheritedElement lastModel = models.last;
+    for (InheritedElement model in models) {
+      final T value = context.inheritFromElement(model, aspect: aspect);
+      if (model == lastModel)
+        return value;
+    }
+    assert(false, 'Unreachable');
+    return null;
+  }
+}
+
+/// An [Element] that uses a [InheritedModel] as its configuration.
+class InheritedModelElement<T> extends InheritedElement {
+  /// Creates an element that uses the given widget as its configuration.
+  InheritedModelElement(InheritedModel<T> widget) : super(widget);
+
+  @override
+  InheritedModel<T> get widget => super.widget;
+
+  @override
+  void updateDependencies(Element dependent, Object aspect) {
+    final Set<T> dependencies = getDependencies(dependent);
+    if (dependencies != null && dependencies.isEmpty)
+      return;
+
+    if (aspect == null) {
+      setDependencies(dependent, new HashSet<T>());
+    } else {
+      assert(aspect is T);
+      setDependencies(dependent, dependencies == null
+        ? (new HashSet<T>()..add(aspect))
+        : dependencies..add(aspect));
```

Comment:
this might be cleaner as:

`(dependencies ?? new HashSet<T>())..add(aspect)`

Reply:
Yes, that's tidier